File: 03_Arrays/Easy/Python/06_move_zeros.py
def move_zeros_brute(arr: list) -> None:

  temp = []

  # store the non-zero elements in an array/list
  for i in range(len(arr)):
    if arr[i] != 0:
      temp.append(arr[i])

  # store that elements in OG array
  for i in range(len(temp)):
    arr[i] = temp[i]


  # rest of the elements in the array will be zero 
  for i in range(len(temp), len(arr)):
    arr[i] = 0

  return


def move_zeros_optimal(arr: list) -> None:

  # j will always point on the zero
  j = None

  # finding index of 1st zero:
  for i in range(len(arr)):
    if arr[i] == 0:
      j = i
      break

  # if no zero elements are found:
  if j == None:
    return
  
  # if zero elements are found:
  for i in range(j+1, len(arr)):

    # if the element is non-zero
    # swap it with j
    if arr[i] != 0:
      arr[i], arr[j] = arr[j], arr[i]
      j += 1
  
  return

def move_zeros_optimal_2(arr: list) -> None:
    # Initialize j to 0 to track the next zero position
    j = 0

    # Iterate through the array
    for i in range(len(arr)):
        # If the element is non-zero,
          # swap it with the element at index j
        if arr[i] != 0:
            arr[i], arr[j] = arr[j], arr[i]
            j += 1

    return


size = int(input("Enter size: "))
arr = list(map(int, input("Enter elements in array: ").strip().split()))

# move_zeros_brute(arr)
# move_zeros_optimal(arr)
move_zeros_optimal_2(arr)

print(arr)

File: 03_Arrays/Easy/C++/06_move_zeros.cpp
#include <bits/stdc++.h>
using namespace std;

void moveZeroes_brute(vector<int>& nums) {
    int n = nums.size();

    vector<int> temp;

    for (int i=0; i<n; ++i){
      if(nums[i] != 0)
        temp.push_back(nums[i]);
    }

    for (int i=0; i<temp.size(); ++i){
      nums[i] = temp[i];
    }

    for (int i=temp.size(); i<n; ++i){
      nums[i] = 0;
    }
}

void mySwap(int &a, int &b){
  int temp = a;
  a = b;
  b = temp;
}
void moveZeroes_Optimal(vector<int>& nums) {
    int j = -1;

    // finding index of 1st zero
    for (int i = 0; i<nums.size(); ++i){
      if(nums[i] == 0){
        j = i;
        break;
      }
    }

    // if no zeros are found:
    if (j == -1) return;

    //Move the pointers i and j
    //and swap accordingly:
        // j+1 -> bcz we don't want to swap any prev non-zero elements
    for (int i = j+1; i<nums.size(); ++i){
      if (nums[i] != 0){
        mySwap(nums[i], nums[j]);
        j++;
      }
    }
}

int main(int argc, char* argv[]){
  vector<int> v;

  int size;
  cin >> size;

  for (int i=0; i<size; ++i) {
    int a;
    cin >> a;
    v.push_back(a); 
  }

  // moveZeroes_brute(v);
  moveZeroes_Optimal(v);

  for (int i=0; i<size; ++i){
    cout << v[i] << " ";
  }
return 0;
}